Adding `Data.Fin.Permutation`

bch29 opened the pull request:
I know this needs some fixing up. One thing to mention: you said to move `swapFin` to `Data.Fin`. However, it relies on `_≟_` which is defined in `Data.Fin.Properties`. Perhaps `_≟_` should be moved to `Data.Fin`?

MatthewDaggitt commented:
> One thing to mention: you said to move swapFin to Data.Fin. However, it relies on _≟_ which is defined in Data.Fin.Properties. Perhaps _≟_ should be moved to Data.Fin?

Yup. Sadly that particular little niggle is my fault. When I added it, I put in the wrong place (okay, a place inconsistent with the rest of the library). I've just fixed it in a mostly backwards compatible way in cedbf87 so feel free to have a go at tidying up the PR.

I'm on holiday for the next week and a half so I may not be very active, but when I'll be on this when I get back.

bch29 commented:
OK, I've moved `swapFin` and done some other cleaning up.

As per @gallais' suggestion I generalized the `if-dec-{true,false}` lemmas and called them `diag` and `antidiag` respectively. Since they now don't mention `Fin`, I've moved them to `Relation.Nullary.Decidable`. As it turns out, they're pretty similar to the `fromWitness` family of functions that is already there, but it would be enough hassle to make use of those that I think it's still worth including my new definitions.

MatthewDaggitt commented:
> As per @gallais' suggestion I generalized the if-dec-{true,false} lemmas and called them diag and antidiag respectively. Since they now don't mention Fin, I've moved them to Relation.Nullary.Decidable. As it turns out, they're pretty similar to the fromWitness family of functions that is already there, but it would be enough hassle to make use of those that I think it's still worth including my new definitions.

You probably won't need any of these proofs if you use patten matching directly in the definition of `swap`:
``` agda
swap : ∀ {n} → Fin n → Fin n → Fin n → Fin n
swap i j k with k ≟ i
... | yes _ = j
... | no  _ with k ≟ j
...   | yes _ = i
...   | no  _ = k
```
This form is much more in keeping with the style of the standard library as it preserves type information (the library tries to avoid `Bool` where possible as it requires all sorts of the lemmas like `diag` and `antidiag` to recover the type information)

gallais commented:
Yeah it's a fairly nasty one. Quick and dirty:

```agda
module Swap where

open import Data.Empty
open import Data.Fin
open import Data.Fin.Properties
open import Data.Product hiding (swap)
open import Function
open import Relation.Nullary
open import Relation.Binary.PropositionalEquality

swap : ∀ {n} → Fin n → Fin n → Fin n → Fin n
swap i j k with k ≟ i
... | yes _ = j
... | no  _ with k ≟ j
...   | yes _ = i
...   | no  _ = k

≡-≟-identity : ∀ {n} (i j : Fin n) (p : i ≡ j) → (i ≟ j) ≡ yes p
≡-≟-identity i j eq with i ≟ j
... | yes p = cong yes (≡-irrelevance p eq)
... | no ¬p = ⊥-elim (¬p eq)

≢-≟-identity : ∀ {n} (i j : Fin n) → ¬ (i ≡ j) → ∃ λ ¬p → (i ≟ j) ≡ no ¬p
≢-≟-identity i j ¬eq with i ≟ j
... | yes p = ⊥-elim (¬eq p)
... | no ¬p = ¬p , refl

swap-inverse : ∀ {n} (i j : Fin n) {k} → swap i j (swap j i k) ≡ k
swap-inverse i j {k} with k ≟ j
... | yes p rewrite ≡-≟-identity i i refl = sym p
... | no ¬p with k ≟ i
swap-inverse i j {k} | no ¬p | yes q with j ≟ i
... | yes r = trans r (sym q)
... | no ¬r rewrite ≡-≟-identity j j refl = sym q
swap-inverse i j {k} | no ¬p | no ¬q
  rewrite proj₂ (≢-≟-identity k i ¬q)
        | proj₂ (≢-≟-identity k j ¬p) = refl
```

bch29 commented:
I had something like that to start with, but `≡-≟-identity` and `≢-≟-identity` are included in "lemmas like `diag` and `antidiag`"! They're just uglier versions that have to account for proof relevance. If `swap` is defined in terms of `Bool`s you get to use the prettier `diag` and `antidiag`.

bch29 commented:
@MatthewDaggitt @gallais it would be nice to move forward with this either way so that I can continue with the rest of #211. I see four options, and I'm happy for you to choose which is made final:

- Keep the definition of `swap` and proof of `swap-inverse` as it is now, using `if_then_else_` and `diag` and `antidiag`.
- Keep the definition of `swap` as is, using `if_then_else_`, but move `diag` and `antidiag` to local definitions in a `where` clause.
- Change `swap` to use pattern matching on the `Dec` value directly, and use the large, explicit proof I wrote out above.
- Change `swap` to use pattern matching on the `Dec` value directly, and use @gallais' proof (but move `≡-≟-identity` and `≢-≟-identity` to local definitions in a `where` clause).

gallais commented:
Btw #266 is a suggestion to add generic versions of `≡-≟-identity` and `≢-≟-identity`
to `Relation.Binary.PropositionalEquality` so they wouldn't need to be defined in a
`where` block.

I don't have any strong opinion one way or the other (except against the big monolithic
proof).

MatthewDaggitt commented:
I think the best thing is to go with @gallais's proof. A few things to do as well before merging in:

- Please could you move all this stuff (minus the `punchIn` and `punchOut` stuff) to a new module called `Data.Fin.Permutation`?
- Move `swap-inverse` out of the `where` block so it's more easily accessible.
- Get rid of all the `ℕ.` as they're not needed.
- Get rid of all the `P.` as propositional equality is the only relation we're using.
- The contents of the `where` block of `removeIn↔` is indented once too many times
- Use superscript and subscript indices for the permutations in `removeIn↔` (i.e. `πʳ` rather than `πR`)
- I don't think there's any need for `πi`, `πʳ i` is much clearer.
- Put the arrows at the end of the lines of the type signatures of `punchIn-permute` and `punchIn-permute′`, and align the type signatures.
- Open `≡-Reasoning` at the top of the file.

MatthewDaggitt commented:
Sorry, I should have been clearer. Could you move `swap` to `.Permutations` as well? It's definitely a permutation operator, rather than an every day operation on `Fin`.

Then I promise we can merge in!

bch29 commented:
If I'm moving `swap` to `.Permutations` then maybe the function itself and `swap-inverse` should be private, and `swap-perm` should be renamed to `swap`. Then you'd recover the original as `Inverse.to swap ⟨$⟩_`. It doesn't really make sense to me to expose half of a permutation separately in a file that exposes the whole permutation anyway.

As another point of discussion, in my previous commit I renamed `punchInᵢ≢i` to `i≢punchInᵢ`, and swapped the inequality around accordingly in the result. I did this thinking I wrote that function in the first place, but I now realise it was already there, so this would be a breaking change. The reason for the swap is that I was composing it with `sym` in every single place I used it, so I thought it would just make more sense this way round. However, if you don't like the breaking change I can swap it back.

bch29 commented:
I also think that now these things are in their own module, it might make sense to rename `removeIn↔` to something like `removeMember`. The logic is that it removes a member of a permutation, which seems more descriptive that the existing name. When using the module, you'd probably import it qualified as `Perm` or similar, so it'd be used as `Perm.removeMember`.

Additionally, would it make sense to include some type synonyms, and additional definitions to make the use of permutations easier (without having to go through `Function.Inverse`)?

```agda
Permutation′ : ℕ → ℕ → Set
Permutation′ m n = Fin m ↔ Fin n

Permutation : ℕ → Set
Permutation n = Permutation′ n n

_⟨$⟩ʳ_ : ∀ {m n} → Permutation′ m n → Fin m → Fin n
_⟨$⟩ʳ_ = _⟨$⟩_ ∘ Inverse.to

_⟨$⟩ˡ_ : ∀ {m n} → Permutation′ m n → Fin n → Fin m
_⟨$⟩ˡ_ = _⟨$⟩_ ∘ Inverse.from

inverseˡ : ∀ {m n} (π : Permutation′ m n) {i} → π ⟨$⟩ˡ (π ⟨$⟩ʳ i) ≡ i
inverseˡ π = Inverse.left-inverse-of π _

inverseʳ : ∀ {m n} (π : Permutation′ m n) {i} → π ⟨$⟩ʳ (π ⟨$⟩ˡ i) ≡ i
inverseʳ π = Inverse.right-inverse-of π _
```

bch29 commented:
Here's what the file looks like with those changes: https://github.com/bch29/agda-stdlib/blob/fin-properties-alt/src/Data/Fin/Permutation.agda

I think it makes a lot of the proofs lower down look quite a bit clearer.

MatthewDaggitt commented:
> If I'm moving swap to .Permutations then maybe the function itself and swap-inverse should be private, and swap-perm should be renamed to swap. Then you'd recover the original as Inverse.to swap ⟨$⟩_. It doesn't really make sense to me to expose half of a permutation separately in a file that exposes the whole permutation anyway.

Generally we like to expose as much of the code as possible. Who knows what uses other people may find for the functions. It's particularly irritating to find something useful in the standard library only to find out that it's private or in a `where` block for no particular reason.

> As another point of discussion, in my previous commit I renamed punchInᵢ≢i to i≢punchInᵢ, and swapped the inequality around accordingly in the result. I did this thinking I wrote that function in the first place, but I now realise it was already there, so this would be a breaking change.

Yes, can you change it back please, we try to minimise any unnecessary non-backwards compatibility.

> I also think that now these things are in their own module, it might make sense to rename removeIn↔ to something like removeMember. The logic is that it removes a member of a permutation, which seems more descriptive that the existing name. When using the module, you'd probably import it qualified as Perm or similar, so it'd be used as Perm.removeMember.

Yup, that sounds fine.

> Additionally, would it make sense to include some type synonyms, and additional definitions to make the use of permutations easier (without having to go through `Function.Inverse`).

Yes, great additions!

> We might consider adding a combinator that shows that Permutation′ is really the same as Permutation, too.

Haven't thought about it much, but do you need `Permutation′`?  If yes, then yes that would be a good lemma to have.

bch29 commented:
> Generally we like to expose as much of the code as possible. Who knows what uses other people may find for the functions.

The original `swap` function is just a matter of applying the permutation in one direction and `swap-inverse i j` can be recovered easily as `inverseʳ (swap i j)`. It doesn't seem to me that you gain anything by exposing the underlying definitions directly, other than adding extra things into the namespace...

> Haven't thought about it much, but do you need `Permutation′`?

Yeah, it's much easier to prove permutations if you don't have to worry about proving that the sizes of the corresponding sets are the same size at the same time.

JacquesCarette commented:
+1000 on the advantages of working with `Permutation′` over `Permutation` !

bch29 commented:
@gallais I'm using `HEAD` as of around 2 weeks ago. I'll try compiling it with 2.5.4 and see if I can see what's going on.

gallais commented:
Strange. I'd have thought you'd need the `experimental` branch of
agda-stdlib for `2.5.4` and above. `master` is only checked with `2.5.3`

bch29 commented:
I don't think most of the library breaks with newer Agda versions, it's mostly the reflection stuff that changes as far as I can tell.

bch29 commented:
Those unsolved metas should be fixed now. This doesn't seem like a regression since the newer version of Agda just able to infer more.

MatthewDaggitt commented:
> The original swap function is just a matter of applying the permutation in one direction and swap-inverse i j can be recovered easily as inverseʳ (swap i j). It doesn't seem to me that you gain anything by exposing the underlying definitions directly, **other than adding extra things into the namespace**...

"Adding extra things to the namespace" is exactly the point! Imagine a situation where someone wants to use `swap` for something other than building a permutation. Why would they want to have to extract it manually from the bijection proof? 

The same argument used to hold sway for lots of other record types. For example, before v0.14 there wasn't an easily accessible proof for associativity of `_+_` over natural numbers. Instead you had to import the `Semiring` proof and open that record extracting `+-assoc`. Doable yes, but resulted in a lot of extra code bloat. Also from an efficiency point of view you don't want to have to build the whole semiring/permutation proof just to extract the one bit of it you're interested in.

> +1000 on the advantages of working with Permutation′ over Permutation !

Okay, I get the message :smile: Anyone have a better name for it than `Permutation′` or are people happy with it?

bch29 commented:
@MatthewDaggitt OK, this all makes sense. My only remaining reservation is that I like the idea of just calling the permutation itself `swap`, because anything else like the old  name `swap-perm` feels awkward in a module that's all about permutations! This could be solved by calling the underlying function something other than `swap`, but I can't come up with a better alternative name at the moment. FWIW I don't agree with the efficiency argument, because as far as I'm aware Agda doesn't evaluate parts of expressions that aren't used.

gallais commented:
> Anyone have a better name for it than `Permutation′` or are people happy with it?

Looking at `Function`, we have tended to use a tick for the *less general* version of a
combinator so I suppose I would expect `Permutation` and `Permutation'` to be the other
way around.

As for the special case where `m` equals `n`, I would jokingly suggest `\MfS` aka `𝔖`
because `𝔖ₙ` is a common notation for the [symmetric group of degree n](https://en.wikipedia.org/wiki/Symmetric_group#Definition_and_first_properties).

gallais commented:
> This could be solved by calling the underlying function something other than swap, but I can't come up with a better alternative name at the moment. 

This function is typically called a [transposition](https://en.wikipedia.org/wiki/Cyclic_permutation#Transpositions) so I suppose `transpose` could do the trick.

MatthewDaggitt commented:
> Looking at Function, we have tended to use a tick for the less general version of a
combinator so I suppose I would expect Permutation and Permutation' to be the other
way around.

Good point!

> This function is typically called a transposition so I suppose transpose could do the trick.

I back the change from `swap` to `transposition`. Much clearer what's going on!

> My only remaining reservation is that I like the idea of just calling the permutation itself swap, because anything else like the old name swap-perm feels awkward in a module that's all about permutations!

Except that I'd argue that the function itself "is" the permutation, and what we're calling a permutation is the proof that it's a permutation and the convention in the existing library is for the functions to have the original names and for proofs about them to be called something else. How about either `transposeₚ` or `transpose↔` for the name of the permutation?

Sorry this taking a while to decide. I promise we will get it merged in!

bch29 commented:
Hopefully this (or a solution like this) leaves everyone happy. The base functions have been moved to `Data.Fin.Permutation.Components`, allowing me to keep the name `transpose` (and also `reverse` now) in both places.

bch29 commented:
An alternative would be to leave everything in the same module and call them verb/noun forms of the same word:
- `reverse/reversal`
- `tranpose/transposition`
- etc

I do have reservations about exposing all of these names in the same file, however. It would be nice to keep `Data.Fin.Permutation` as a pure permutation-reasoning library.

Now that I put it that way, perhaps it should re-export equational reasoning combinators from `Function.Related` too...

MatthewDaggitt commented:
> Hopefully this (or a solution like this) leaves everyone happy. The base functions have been moved to Data.Fin.Permutation.Components, allowing me to keep the name transpose (and also reverse now) in both places.

Excellent compromise, thanks. The new definitions look really neat!

> Now that I put it that way, perhaps it should re-export equational reasoning combinators from Function.Related too...

I think people can do that themselves. Unless you can think of some compelling renaming of the combinators that only applies to permutations?

Otherwise I'm happy to merge this in. Anyone have any more comments?

MatthewDaggitt commented:
Oh hang on the Travis checks are failing. @bch29 please run "make test" and fix the bugs caused by moving `reverse`, and please document the changes in the CHANGELOG

gallais commented on src/Data/Fin/Permutation/Components.agda:
```diff
@@ -19,8 +19,15 @@ open import Algebra.FunctionProperties using (Involutive)
+-- swap i j i ≡ j
+-- swap i j j ≡ i
+-- swap i j k ≡ k if k ≠ i and k ≠ j
+
```
Should the comment also be updated to use the new name?

bch29 replied:
Done

bch29 commented:
@MatthewDaggitt All done.

MatthewDaggitt commented:
Okay thanks. I'll leave it up for a couple of days to see if anyone else has any comments and then I'll merge it in.
